rootfs-builder/clearlinux/rootfs_lib.sh

```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+#
+# Copyright (c) 2017 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+set -e
+
+
+
+
+check_program(){
+	type "$1" >/dev/null 2>&1
+}
+
+generate_dnf_config()
+{
+	cat > "${DNF_CONF}" << EOF
+[main]
+cachedir=/var/cache/dnf/clear/
+keepcache=0
+debuglevel=2
+logfile=/var/log/dnf.log
+exactarch=1
+obsoletes=1
+gpgcheck=0
+plugins=0
+installonly_limit=3
+#Dont use the default dnf reposdir
+#this will prevent to use host repositories
+reposdir=/root/mash
+
+[clear]
+name=Clear
+failovermethod=priority
+baseurl=${REPO_URL}
+enabled=1
+gpgcheck=0
+EOF
+}
+
+build_rootfs()
+{
+	local ROOTFS_DIR=$1
+	local REPO_URL=${REPO_URL:-https://download.clearlinux.org/current/x86_64/os/}
+	local EXTRA_PKGS=${EXTRA_PKGS:-}
+	check_root
+	if [ ! -f "${DNF_CONF}" ]; then
+		DNF_CONF="./clear-dnf.conf"
+		generate_dnf_config
+	fi
+	mkdir -p "${ROOTFS_DIR}"
+	if [ -n "${PKG_MANAGER}" ]; then
+		info "DNF path provided by user: ${PKG_MANAGER}"
+	elif check_program "dnf"; then
+		PKG_MANAGER="dnf"
+	elif check_program "yum" ; then
+		PKG_MANAGER="yum"
+	else
+		die "neither yum nor dnf is installed"
+	fi
+
+	info "Using : ${PKG_MANAGER} to pull packages from ${REPO_URL}"
+
+	DNF="${PKG_MANAGER} --config=$DNF_CONF -y --installroot=${ROOTFS_DIR} --noplugins"
+	$DNF install ${EXTRA_PKGS} ${PACKAGES}
+
+	[ -n "${ROOTFS_DIR}" ]  && rm -r "${ROOTFS_DIR}/var/cache/dnf"
+}
+
+
+check_root()
```

Comment:
There are multiple instances of `check_program` and `check_root`. How about:

- creating a minimal `utils.sh` for these functions.
- making the existing script source `utils.sh`.
- adding a comment in the doc stating that `utils.sh` might be worth sourcing (in fact I'd just do it in the template scripts).